neps/nep-504.md

```diff
@@ -0,0 +1,242 @@
+    "telegram": ["telegram_handle_1", "telegram_handle2"], // must be a list
+  },
+  // Default token icon
+  "icon": {
+    "png": "https://genadrop.mypinata.cloud/ipfs/...",
+    "svg": "https://genadrop.mypinata.cloud/ipfs/..."
+  },
+  // additional video content. Sub object, where key is a video format.
+  "video": {
+    "mp4": "https://youtube.com/...",
+    "vp9": "ipfs://..."
+  },
+  // additional audio content. Sub object, where key is an audio format.
+  "audio": {
+    "ogg": "ipfs://..."
+  }
+```
+
+All reference fields (`TokenMetadata.reference`, `ClassMetadata.reference`, `ContractMetadata.reference`) share the same scheme. Common information should be pushed up in the category. For example project information should be stored in Class Metadata or Contract Metadata, rather than the Token Metadata. Issuer information and issuer website should be stored in the `ContractMetadata`.
+
+### Information flow
+
+Projects and dapps should use all stack of information to correctly render token information. Let's consider the following examples, where each token is part of the same class (`class1`) and same issuer (`issuer1`)
+
+1. All `token1metadata.reference.icon`, `class1metadata.reference.icon` and `issuer1metadata.reference.icon` are defined. To render token, the `token1metadata` icon field should be used. Moreover, each application can choose which format (png, jpg, svg...) to use.
+1. All `token1metadata.reference.icon` is not defined and `class1metadata.reference.icon` is defined. The `class1metadata` icon should be used to render token.
+1. Both `token1metadata.reference.webiste` and `class1metadata.reference.webiste` are defined. A dapp can use both token and class level metadata to provide detailed information.
+
+### Events
+
+Whenever a metadata is updated, an event should be emitted. We propose to follow the same event principles as defined in NEP-393:
+
+- Events don't need to repeat all function arguments - these are easy to retrieve by indexer (events are consumed by indexers anyway).
+- Events must include fields necessary to identify subject matters related to use case.
+- When possible, events should contain aggregated data, with respect to the standard function related to the event.
+
+The Contract and Class Metadata are stored and managed by the Issuer contract. This NEP covers only the Issuer contract. `TokenMetadata` is stored in the SBT registry, consequently Token Metadata updates and events are defined in NEP-393.
+
+```typescript
+type Nep504Event {
+  standard: "nep504";
+  version: "1.0.0";
+  event: "class" | "contract";
+  data: Class | Contract ;
+}
+
+/// An event emitted by an SBT Issuer when Class Metadata are updated.
+type Class {
+  classes: ClassId[];  // list of class IDs whose metadata has been updated
+}
+
+/// An event emitted by an SBT Issuer when Contract Metadata is updated.
+/// Note: it's an empty object
+type Class { }
```

Comment:
Did you mean `Contract`?
I would disambiguate event types: maybe `ClassData` and `ContractData` would sound better (?)